src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs

```diff
@@ -0,0 +1,115 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Playwright.TestAdapter;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Playwright.MSTest;
+
+internal class PlaywrightTimeoutAttribute : TestMethodAttribute
+{
+    static internal Func<Task>? ContextCloseHookOnTimeout { get; set; } = null;
+
+    public override TestResult[] Execute(ITestMethod testMethod)
+    {
+        var timeout = PlaywrightSettingsProvider.TestTimeout;
+        if (timeout == null || Debugger.IsAttached)
+        {
+            return base.Execute(testMethod);
+        }
+        if (CheckThatPlaywrightTimeoutIsSmallerThanNUnitTimeout(testMethod, timeout.Value) is var result && result != null)
+        {
+            return result;
+        }
+
+        return ExecuteWithTimeoutAsync(testMethod, timeout.Value).GetAwaiter().GetResult();
+    }
+
+    private async Task<TestResult[]> ExecuteWithTimeoutAsync(ITestMethod testMethod, int timeout)
+    {
+        // 1. Race Test execution against the specified timeout.
+        // In MSTest base.Execute will internally run the TestInitialize and TestCleanup. In case of a timeout, we don't
+        // run TestCleanup, only TestInitialize. To get good error messages, we close the browser context inside ContextCloseHookOnTimeout.
+        var testExecutionTask = Task.Run(() => base.Execute(testMethod));
+        var timeoutTask = Task.Delay(timeout);
+        await Task.WhenAny(testExecutionTask, timeoutTask);
```

Comment:
This returns the task that completed that you don't capture. I am not sure if you don't use it on purpose because it made the code read worse, or if you don't know about it. 

That said I don't see any problem with checking the `.IsCompleted` on the task. The worst thing that I can see happen is that you await the test again in 3. if the timeout and test both completed because ContextCloseHookOnTimeout took enough time for both of them to complete. Wrapping a completed task into another task and awaiting it again has not functional issue, other than a tiny tiny added overhead.